model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java

```diff
@@ -202,38 +194,32 @@ protected void registerClusterListeners(KeycloakSession session) {
         KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
         ClusterProvider cluster = session.getProvider(ClusterProvider.class);
 
-        cluster.registerListener(REALM_REMOVED_SESSION_EVENT, new AbstractUserSessionClusterListener<RealmRemovedSessionEvent>(sessionFactory) {
+        cluster.registerListener(REALM_REMOVED_SESSION_EVENT,
+                new AbstractUserSessionClusterListener<RealmRemovedSessionEvent, UserSessionProvider>(sessionFactory, UserSessionProvider.class) {
 
             @Override
-            protected void eventReceived(KeycloakSession session, InfinispanUserSessionProvider provider, RealmRemovedSessionEvent sessionEvent) {
-                provider.onRealmRemovedEvent(sessionEvent.getRealmId());
+            protected void eventReceived(KeycloakSession session, UserSessionProvider provider, RealmRemovedSessionEvent sessionEvent) {
+                ((InfinispanUserSessionProvider) provider).onRealmRemovedEvent(sessionEvent.getRealmId());
```

Comment:
Is it certain that this event has been sent from `InfinispanUserSessionProvider`? Could it have originated in a `MapUserSessionProvider`?

Reply:
Initially my understanding was that the event can come only from `InfinispanUserSessionProvider` because currently we don't use both (map, infinispan) providers simultaneously. After your comment I added a guard to check type of `UserSessionProvider`. Let me know if you have a different solution on mind.

Reply:
This works for this PR, as there is now only a single such a provider active thorough the server lifespan.
It will require change later on, once there would be possibility to use two at the same time